lib/one_wire/ibutton/protocols/dallas/protocol_ds1971.c

```diff
@@ -0,0 +1,258 @@
+#define DS1971_CMD_COPY_SCRATCHPAD 0x55
+#define DS1971_CMD_WRITE_SCRATCHPAD 0x0F
+#define DS1971_CMD_FINALIZATION 0xA5
+
+typedef struct {
+    OneWireSlave* bus;
+    DallasCommonCommandState command_state;
+} DS1971ProtocolState;
+
+typedef struct {
+    DallasCommonRomData rom_data;
+    uint8_t sram_data[DS1971_SRAM_DATA_SIZE];
+    DS1971ProtocolState state;
+} DS1971ProtocolData;
+
+static bool dallas_ds1971_read(OneWireHost*, void*);
+static bool dallas_ds1971_write_copy(OneWireHost*, iButtonProtocolData*);
+static void dallas_ds1971_emulate(OneWireSlave*, iButtonProtocolData*);
+static bool dallas_ds1971_load(FlipperFormat*, uint32_t, iButtonProtocolData*);
+static bool dallas_ds1971_save(FlipperFormat*, const iButtonProtocolData*);
+static void dallas_ds1971_render_data(FuriString*, const iButtonProtocolData*);
+static void dallas_ds1971_render_brief_data(FuriString*, const iButtonProtocolData*);
+static void dallas_ds1971_render_error(FuriString*, const iButtonProtocolData*);
+static bool dallas_ds1971_is_data_valid(const iButtonProtocolData*);
+static void dallas_ds1971_get_editable_data(iButtonEditableData*, iButtonProtocolData*);
+static void dallas_ds1971_apply_edits(iButtonProtocolData*);
+static bool
+    dallas_ds1971_read_mem(OneWireHost* host, uint8_t address, uint8_t* data, size_t data_size);
+
+const iButtonProtocolDallasBase ibutton_protocol_ds1971 = {
+    .family_code = DS1971_FAMILY_CODE,
+    .features = iButtonProtocolFeatureExtData | iButtonProtocolFeatureWriteCopy,
+    .data_size = sizeof(DS1971ProtocolData),
+    .manufacturer = DALLAS_COMMON_MANUFACTURER_NAME,
+    .name = DS1971_FAMILY_NAME,
+
+    .read = dallas_ds1971_read,
+    .write_blank = NULL, /* No data to write a blank */
+    .write_copy = dallas_ds1971_write_copy,
+    .emulate = dallas_ds1971_emulate,
+    .save = dallas_ds1971_save,
+    .load = dallas_ds1971_load,
+    .render_data = dallas_ds1971_render_data,
+    .render_brief_data = dallas_ds1971_render_brief_data,
+    .render_error = dallas_ds1971_render_error,
+    .is_valid = dallas_ds1971_is_data_valid,
+    .get_editable_data = dallas_ds1971_get_editable_data,
+    .apply_edits = dallas_ds1971_apply_edits,
+};
+
+bool dallas_ds1971_read(OneWireHost* host, iButtonProtocolData* protocol_data) {
+    DS1971ProtocolData* data = protocol_data;
+    return onewire_host_reset(host) && dallas_common_read_rom(host, &data->rom_data) &&
+           dallas_ds1971_read_mem(host, 0, data->sram_data, DS1971_SRAM_DATA_SIZE);
+}
+
+bool dallas_ds1971_write_copy(OneWireHost* host, iButtonProtocolData* protocol_data) {
+    DS1971ProtocolData* data = protocol_data;
+
+    onewire_host_reset(host);
+    onewire_host_skip(host);
+    // Starting writing from address 0x0000
+    onewire_host_write(host, DS1971_CMD_WRITE_SCRATCHPAD);
+    onewire_host_write(host, 0x00);
+    // Write data to scratchpad
+    onewire_host_write_bytes(host, data->sram_data, DS1971_SRAM_DATA_SIZE);
+    furi_delay_ms(250);
```

Comment:
`DS1971_COPY_SCRATCH_TIMEOUT_US` is defined, but not used anywhere. However, there are several mentions of 250 milli_seconds here. Please double-check whether the specification calls for 250 micro_ or milli_seconds and use a `#define` constant.